edu.cuny.hunter.streamrefactoring.core/src/edu/cuny/hunter/streamrefactoring/core/analysis/Util.java

```diff
@@ -407,13 +409,38 @@ static int getLineNumberFromIR(IBytecodeMethod method, SSAInstruction instructio
 		return lineNumberFromIR;
 	}
 
+	/**
+	 * This set is used to store visited values. Without this set, the method
+	 * getPossibleTypes could be infinitely recursively called and would never
+	 * return. So adding this set means to add a base case.
+	 */
+	static private HashSet<Value> seenValues = new HashSet();
+
+	static Collection<TypeAbstraction> getPossibleStreamTypes(int valueNumber, TypeInference inference) {
+		seenValues.clear();
```

Comment:
Why is the `clear()` only here?

Reply:
Each time call `getPossibleTypes`, this set must be cleared. The reason that infinite recursion happens here is that if the value is `PhiValue`, the `Uses` could be visited by the method before. 
For example:
No. 1 call:
**Value 3 -> Uses 5 7**
No. 2 call:
Value 5 -> return type
**Value 7** -> Uses 4 **3**
No. 3 call:
**Value 3 -> Uses 5 7**
Becase 3 is visited in the  No. 1 call, so the No. 3 call should be quited. That's why I added a set to store the visited values.

Reply:
Why is it *only* here, meaning that there are other places that may need to be considered (like the other overloaded methods).

Reply:
Done.